[MCC-238763] add self registration to sample

kbohlmann-mdsol opened the pull request:
Self registration added to Android AppConnect sample app. 

Fairly straightforward interactive wizard, a single activity per question, successful registration unwinds this stack back to the login screen and populates the login field, and clears the password field. 

A few inconsistencies from the ios app : 
 we fetch the security questions via appconnect api, in ios they are hardcoded. (we should fix this is in ios)
 we pre validate the fields with specific requirements, rather than raise an alert for incorrect entries, this applies to email entry, password entry, and security question answer entry.

We don't use the isRegistrationActive API call in either sample app, not sure if we want to add that.

This should be good to go, pending testing registration with a valid sandbox token I'll remove the dnm label.

kbohlmann-mdsol committed:
added initial skeleton for self registration

kbohlmann-mdsol committed:
fixed a few copy pastas in initial skeleton

kbohlmann-mdsol committed:
more registration activity skeleton , interim

kbohlmann-mdsol committed:
prevalidate submit button enabling on email and password
cleaned up skeleton flow, finished email and password.

kbohlmann-mdsol committed:
finished gui registration flow

kbohlmann-mdsol committed:
small changes after network build out

kbohlmann-mdsol committed:
merged release into branch

ncarroll-mdsol commented on app/src/main/java/com/sample/appconnectsample/LoginActivity.java:
```diff
@@ -53,6 +53,22 @@ public void doLogInButton(View source) {
+                usernameField.setText(data.getStringExtra("email"));
+                passwordField.setText("");
+            }
+        }
```
Nitpick: nested ifs is probably overkill for two conditions. Could combine them or return early.

ncarroll-mdsol commented on app/src/main/java/com/sample/appconnectsample/RegistrationEmailActivity.java:
```diff
@@ -0,0 +1,73 @@
+            @Override
+            public void afterTextChanged(Editable s) {
+                validate();
+            }
```
Can we use the same listener for both the email field and confirmation field?

ncarroll-mdsol commented on app/src/main/java/com/sample/appconnectsample/RegistrationEmailActivity.java:
```diff
@@ -0,0 +1,73 @@
+        submitButton.setEnabled(false);
+
+        if (emailField.getText().toString().equals(emailConfirmationField.getText().toString()) &&
+                android.util.Patterns.EMAIL_ADDRESS.matcher(emailField.getText().toString()).matches())
```
Nitpick: this could be a bit shorter by extracting the strings to variables.

ncarroll-mdsol commented on app/src/main/java/com/sample/appconnectsample/RegistrationSecurityAnswerActivity.java:
```diff
@@ -0,0 +1,149 @@
+        createAccountButton.setEnabled(false);
+
+        // security answer must be at least 2 characters long.
+        if (securityAnswerField.getText().toString().length() > 2) {
```
Should be `>=` 2 I think?

kbohlmann-mdsol committed:
cleanup of registration sample code

kbohlmann-mdsol committed:
full registration path in place
added network request for security questsions and matched selection question to id from server for submission. added progress spinners for slower network conditions. added client call for registration. no call to isregistrationactive here or in ios, a few other inconsistencies with ios client

kbohlmann-mdsol committed:
moved activity spinner to bottom for registration screen

kbohlmann-mdsol committed:
cleaned up registration activity base handler conditional

kbohlmann-mdsol committed:
made sure not ok results will also filter back through intent

kbohlmann-mdsol committed:
update appconnect comment in sec question fetch

ncarroll-mdsol commented on app/src/main/java/com/sample/appconnectsample/RegistrationSecurityAnswerActivity.java:
```diff
@@ -0,0 +1,138 @@
+
+    private void validate() {
+
+        createAccountButton.setEnabled(false);
```
super nitpick: empty line at the beginning of some functions

kbohlmann-mdsol committed:
cleaned up

kbohlmann-mdsol committed:
updated babbage build number for missing api token

kbohlmann-mdsol committed:
registration working. a few more odds and ends to polish it up.

kbohlmann-mdsol commented on app/src/main/java/com/sample/appconnectsample/App.java:
```diff
@@ -33,11 +33,19 @@ public void onCreate() {
 
+        // It is possible to communicate with other environments in the Medidata
+        // Platform (although in most cases Production should be used).
+        if (BuildConfig.VALIDATION) {
```
Moved this from LoginActivity, think it makes more sense here.

kbohlmann-mdsol commented on app/src/main/java/com/sample/appconnectsample/LoginActivity.java:
```diff
@@ -53,6 +45,28 @@ public void doLogInButton(View source) {
+
+            passwordField.requestFocus();
+
+            new AlertDialog.Builder(LoginActivity.this).
```
Added a popup to inform that registration is successful, and password field is highlighted for entry, email field pre populated.

ncarroll-mdsol commented:
Looks great!
